js/private/npm_import.bzl

```diff
@@ -0,0 +1,248 @@
+
+# TODO: handle npm packages with lifecycle hooks; src will be the output of the postinstall rule instead
+_NODEJS_PACKAGE_TMPL = \
+    """def node_package():
+    if "{link_package_guard}" != "." and native.package_name() != "{link_package_guard}":
+        fail("The node_package() macro loaded from {node_package_bzl} may only be called in the '{link_package_guard}' package. Move the call to the '{link_package_guard}' package BUILD file.")
+
+    _node_package(
+        name = "{namespace}__{bazel_name}",
+        src = "@{rctx_name}//:{dir}",
+        package = "{package}",
+        version = "{version}",
+        visibility = ["//visibility:public"],{maybe_indirect}{maybe_deps}
+    )
+"""
+
+_NODEJS_PACKAGE_EXPERIMENTAL_REF_DEPS_TMPL = _NODEJS_PACKAGE_TMPL + \
+                                             """   _node_package(
+        name = "{namespace}__{bazel_name}__ref",
+        package = "{package}",
+        version = "{version}",{maybe_indirect}
+    )
+"""
+
+_ALIAS_TMPL = \
+    """    native.alias(
+        name = "{namespace}__{alias}",
+        actual = ":{namespace}__{bazel_name}",
+        visibility = ["//visibility:public"],
+    )
+
+    native.alias(
+        name = "{namespace}__{alias}__dir",
+        actual = ":{namespace}__{bazel_name}__dir",
+        visibility = ["//visibility:public"],
+    )
+"""
+
+def _impl(rctx):
+    tarball = "package.tgz"
+    rctx.download(
+        output = tarball,
+        url = "https://registry.npmjs.org/{0}/-/{1}-{2}.tgz".format(
+            rctx.attr.package,
+            # scoped packages contain a slash in the name, which doesn't appear in the later part of the URL
+            rctx.attr.package.split("/")[-1],
+            npm_utils.strip_peer_dep_version(rctx.attr.version),
+        ),
+        integrity = rctx.attr.integrity,
+    )
+
+    dirname = "package"
+    mkdir_args = ["mkdir", "-p", dirname] if not is_windows_os(rctx) else ["cmd", "/c", "if not exist {dir} (mkdir {dir})".format(dir = dirname.replace("/", "\\"))]
+    result = rctx.execute(mkdir_args)
+    if result.return_code:
+        msg = "mkdir %s failed: \nSTDOUT:\n%s\nSTDERR:\n%s" % (dirname, result.stdout, result.stderr)
+        fail(msg)
+
+    # npm packages are always published with one top-level directory inside the tarball, tho the name is not predictable
+    # so we use tar here which takes a --strip-components N argument instead of rctx.download_and_extract
+    untar_args = ["tar", "-xf", tarball, "--strip-components", str(1), "-C", dirname]
+    result = rctx.execute(untar_args)
+    if result.return_code:
+        msg = "tar %s failed: \nSTDOUT:\n%s\nSTDERR:\n%s" % (dirname, result.stdout, result.stderr)
+        fail(msg)
+
+    rctx.file("BUILD.bazel", "exports_files([\"{dir}\"])".format(dir = dirname))
+
+    deps = []
+    for dep in rctx.attr.deps:
+        dep_split = dep.split("@")
+        dep_name = "@".join(dep_split[:-1])
+        dep_version = dep_split[-1]
```

Comment:
This doesn't work in the context of peer dependencies that have multiple `@`s (more than 2)

now:
```
============
Original: @babel/plugin-syntax-class-static-block@7.14.5__babel_core_7.14.3
Name: @babel/plugin-syntax-class-static-block
Version: 7.14.5__babel_core_7.14.3
============
Original: eslint-config-prettier@6.15.0_eslint@7.12.1
Name: eslint-config-prettier@6.15.0_eslint
Version: 7.12.1
```

Should be
```
============
Original: eslint-config-prettier@6.15.0_eslint@7.12.1
Name: eslint-config-prettier
Version: 6.15.0_eslint@7.12.1
============
Original: @typescript-eslint/parser@4.13.0_eslint@7.12.1+typescript@4.5.4
Name: @typescript-eslint/parser
Version: 4.13.0_eslint@7.12.1+typescript@4.5.4
```

Maybe something like:
```
        scope = ""
        if dep.startswith("@"):
            scope = "@"
            dep = dep[1:]
        dep_split = dep.partition("@")
        dep_name = scope + dep_split[0]
        dep_version = dep_split[2]
```

Reply:
Thanks for the suggestion. I've some time today finally to understand the proper shape for peer deps in this rule set. Will reply back here when I find it.

Reply:
Sorted out. The correct shape for peer deps in now landed on main branch.